### PR TITLE
e2e tests

### PR DESCRIPTION
**Pull request by witzig**
Submitting for discussion. 

Uses `mocha`, `selenium-webdriver` and `phantomjs`. The following tests are currently implemented:

```
  e2e
    tests/login
      login
        ✓ can access home page (2112ms)
        ✓ can not access restricted content (340ms)
        ✓ can not login with false credentials (841ms)
        ✓ can login as admin (873ms)
        ✓ can access account page as admin (412ms)
        ✓ can logout (391ms)
    tests/subscription
      subscribe (list one)
        ✓ visits web-subscribe (256ms)
        ✓ submits invalid email (error) (301ms)
        ✓ submits valid email (458ms)
        ✓ sees web-confirm-subscription-notice (40ms)
        ✓ receives mail-confirm-subscription (246ms)
        ✓ clicks confirm subscription (156ms)
        ✓ sees web-subscribed-notice (69ms)
        ✓ receives mail-subscription-confirmed (253ms)
      unsubscribe (list one)
        ✓ clicks unsubscribe (164ms)
        ✓ sees web-unsubscribe
        ✓ clicks confirm unsubscription (158ms)
        ✓ sees web-unsubscribed-notice
        ✓ receives mail-unsubscription-confirmed (145ms)
```

The idea is to test the most critical parts, e.g. the subscription process. I would like to extend it a bit, yet I'm not really sold on the current page objects. See:

- [subscription test file](https://github.com/witzig/mailtrain/blob/6c35046ab2ba3052c1c621b5b88e0ce2eeb967a4/test/e2e/tests/subscription.js)
- [subscription page objects](https://github.com/witzig/mailtrain/blob/6c35046ab2ba3052c1c621b5b88e0ce2eeb967a4/test/e2e/page-objects/subscription.js)
- [base page object](https://github.com/witzig/mailtrain/blob/6c35046ab2ba3052c1c621b5b88e0ce2eeb967a4/test/e2e/page-objects/page.js)

@andris9, @bures - What do you think?

Thanks.

### CODE REVIEW

**Comment by bures** [comment]
Excellent! Thanks a lot. Indeed, I felt like it would be good to have some tests when I was refactoring the subscription mechanism. There are many different use-cases, which are just silly to repeat manually every time one introduces a fix.

I had a quick look on the code. Looks good to me. I will give it a bit more detailed look and will try to execute the tests in the evening or tomorrow.

**Comment by bures** [comment]
Dear Roger. It looks perfect. I just have some difficulties to set it all correctly. 
Please, do you think you could include some shell script which (in similar vein has setup/install.sh does) would setup the database, user and test.toml configuration?

I tried some setup myself, but I'm obviously doing something wrong as I'm getting some errors when running the test server and the e2e test.

Thanks.
T.

**Comment by witzig** [comment]
Tomas, sure thing. Also updated the [readme](https://github.com/witzig/mailtrain/blob/43e1cab315a749042886767de599932335c34c5e/test/e2e/README.md). Hope this helps?!

File: test/e2e/README.md [hunk 1]
```diff
@@ -1,6 +1,44 @@
...
+
+These e2e tests have to be performed against predefined resources (e.g. lists, users, etc.) and therefore a dedicated test database and test config is required.
+
+Both can be created by running `sudo sh test/e2e/install.sh` from within your mailtrain directory. This creates a MYSQL user and database called `mailtrain_test`, and generates the required `config/test.toml`.
+
+## Running e2e Tests
+
+For tests to succeed Mailtrian must be started in `test` mode on port 3000 (as http://localhost:3000/ is the predefined service url). The tests itself have to be started in a second Terminal window.
 
 1. Start Mailtrain with `npm run startest`
```
Comment (bures) [inline test/e2e/README.md]: 'startest' -> 'starttest'

**Comment by bures** [comment]
@witzig - thank you for the install script. Works nicely. 

I'm still getting the errors below. Could it be that something's missing in the DB dump?

Thanks a lot.
T.

```
[root@mail mailtrain]# npm run starttest

> mailtrain@1.23.2 starttest /opt/mailtrain
> NODE_ENV=test node index.js

info Using local auth 
No translation was found for msgid "Email address" in msgctxt "" and domain "en"
No translation was found for msgid "Signup country" in msgctxt "" and domain "en"
No translation was found for msgid "Sign up date" in msgctxt "" and domain "en"
No translation was found for msgid "Latest open" in msgctxt "" and domain "en"
No translation was found for msgid "Latest click" in msgctxt "" and domain "en"
No translation was found for msgid "First name" in msgctxt "" and domain "en"
No translation was found for msgid "Last name" in msgctxt "" and domain "en"
No translation was found for msgid "Text" in msgctxt "" and domain "en"
No translation was found for msgid "Website" in msgctxt "" and domain "en"
No translation was found for msgid "Multi-line text" in msgctxt "" and domain "en"
No translation was found for msgid "GPG Public Key" in msgctxt "" and domain "en"
No translation was found for msgid "Number" in msgctxt "" and domain "en"
No translation was found for msgid "Radio Buttons" in msgctxt "" and domain "en"
No translation was found for msgid "Checkboxes" in msgctxt "" and domain "en"
No translation was found for msgid "Drop Down" in msgctxt "" and domain "en"
No translation was found for msgid "Date (MM/DD/YYY)" in msgctxt "" and domain "en"
No translation was found for msgid "Date (DD/MM/YYYY)" in msgctxt "" and domain "en"
No translation was found for msgid "Birthday (MM/DD)" in msgctxt "" and domain "en"
No translation was found for msgid "Birthday (DD/MM)" in msgctxt "" and domain "en"
No translation was found for msgid "JSON value for custom rendering" in msgctxt "" and domain "en"
No translation was found for msgid "Option" in msgctxt "" and domain "en"
No translation was found for msgid "Sign up date" in msgctxt "" and domain "en"
No translation was found for msgid "Latest open" in msgctxt "" and domain "en"
No translation was found for msgid "Latest click" in msgctxt "" and domain "en"
No translation was found for msgid "Delivered" in msgctxt "" and domain "en"
No translation was found for msgid "Has Opened" in msgctxt "" and domain "en"
No translation was found for msgid "Has Clicked" in msgctxt "" and domain "en"
No translation was found for msgid "Not Opened" in msgctxt "" and domain "en"
No translation was found for msgid "Not Clicked" in msgctxt "" and domain "en"
info sql SQL not set up, initializing
info sql Loading tables from mailtrain-test.sql
info sql Database check completed
info Express WWW server listening on port 3000
info Test SMTP Server listening on port 5587
info Test SMTP Mail Box Server listening on port 3001
verb UTC Updating timezone offsets
info Service All services started
No translation was found for msgid "Text" in msgctxt "" and domain "en"
No translation was found for msgid "Website" in msgctxt "" and domain "en"
No translation was found for msgid "Multi-line text" in msgctxt "" and domain "en"
No translation was found for msgid "GPG Public Key" in msgctxt "" and domain "en"
No translation was found for msgid "Number" in msgctxt "" and domain "en"
No translation was found for msgid "Radio Buttons" in msgctxt "" and domain "en"
No translation was found for msgid "Checkboxes" in msgctxt "" and domain "en"
No translation was found for msgid "Drop Down" in msgctxt "" and domain "en"
No translation was found for msgid "Date (MM/DD/YYY)" in msgctxt "" and domain "en"
No translation was found for msgid "Date (DD/MM/YYYY)" in msgctxt "" and domain "en"
No translation was found for msgid "Birthday (MM/DD)" in msgctxt "" and domain "en"
No translation was found for msgid "Birthday (DD/MM)" in msgctxt "" and domain "en"
No translation was found for msgid "JSON value for custom rendering" in msgctxt "" and domain "en"
No translation was found for msgid "Option" in msgctxt "" and domain "en"
No translation was found for msgid "Email address" in msgctxt "" and domain "en"
No translation was found for msgid "Signup country" in msgctxt "" and domain "en"
No translation was found for msgid "Sign up date" in msgctxt "" and domain "en"
No translation was found for msgid "Latest open" in msgctxt "" and domain "en"
No translation was found for msgid "Latest click" in msgctxt "" and domain "en"
No translation was found for msgid "First name" in msgctxt "" and domain "en"
No translation was found for msgid "Last name" in msgctxt "" and domain "en"
ERR! Triggers Error: ER_NO_SUCH_TABLE: Table 'mailtrain_test.triggers' doesn't exist
ERR! Triggers     at Query.Sequence._packetToError (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Sequence.js:52:14)
ERR! Triggers     at Query.ErrorPacket (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Query.js:77:18)
ERR! Triggers     at Protocol._parsePacket (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:280:23)
ERR! Triggers     at Parser.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Parser.js:75:12)
ERR! Triggers     at Protocol.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:39:16)
ERR! Triggers     at Socket.<anonymous> (/opt/mailtrain/node_modules/mysql/lib/Connection.js:103:28)
ERR! Triggers     at emitOne (events.js:96:13)
ERR! Triggers     at Socket.emit (events.js:188:7)
ERR! Triggers     at readableAddChunk (_stream_readable.js:176:18)
ERR! Triggers     at Socket.Readable.push (_stream_readable.js:134:10)
ERR! Triggers     --------------------
ERR! Triggers     at Protocol._enqueue (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:141:48)
ERR! Triggers     at PoolConnection.query (/opt/mailtrain/node_modules/mysql/lib/Connection.js:208:25)
ERR! Triggers     at db.getConnection (/opt/mailtrain/services/triggers.js:29:20)
ERR! Triggers     at Ping.onOperationComplete [as _callback] (/opt/mailtrain/node_modules/mysql/lib/Pool.js:110:5)
ERR! Triggers     at Ping.Sequence.end (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Sequence.js:86:24)
ERR! Triggers     at Ping.Sequence.OkPacket (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Sequence.js:95:8)
ERR! Triggers     at Protocol._parsePacket (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:280:23)
ERR! Triggers     at Parser.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Parser.js:75:12)
ERR! Triggers     at Protocol.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:39:16)
ERR! Triggers     at Socket.<anonymous> (/opt/mailtrain/node_modules/mysql/lib/Connection.js:103:28)
ERR! Triggers  { Error: ER_NO_SUCH_TABLE: Table 'mailtrain_test.triggers' doesn't exist
ERR! Triggers     at Query.Sequence._packetToError (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Sequence.js:52:14)
ERR! Triggers     at Query.ErrorPacket (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Query.js:77:18)
ERR! Triggers     at Protocol._parsePacket (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:280:23)
ERR! Triggers     at Parser.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Parser.js:75:12)
ERR! Triggers     at Protocol.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:39:16)
ERR! Triggers     at Socket.<anonymous> (/opt/mailtrain/node_modules/mysql/lib/Connection.js:103:28)
ERR! Triggers     at emitOne (events.js:96:13)
ERR! Triggers     at Socket.emit (events.js:188:7)
ERR! Triggers     at readableAddChunk (_stream_readable.js:176:18)
ERR! Triggers     at Socket.Readable.push (_stream_readable.js:134:10)
ERR! Triggers     --------------------
ERR! Triggers     at Protocol._enqueue (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:141:48)
ERR! Triggers     at PoolConnection.query (/opt/mailtrain/node_modules/mysql/lib/Connection.js:208:25)
ERR! Triggers     at db.getConnection (/opt/mailtrain/services/triggers.js:29:20)
ERR! Triggers     at Ping.onOperationComplete [as _callback] (/opt/mailtrain/node_modules/mysql/lib/Pool.js:110:5)
ERR! Triggers     at Ping.Sequence.end (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Sequence.js:86:24)
ERR! Triggers     at Ping.Sequence.OkPacket (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Sequence.js:95:8)
ERR! Triggers     at Protocol._parsePacket (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:280:23)
ERR! Triggers     at Parser.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Parser.js:75:12)
ERR! Triggers     at Protocol.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:39:16)
ERR! Triggers     at Socket.<anonymous> (/opt/mailtrain/node_modules/mysql/lib/Connection.js:103:28)
ERR! Triggers   code: 'ER_NO_SUCH_TABLE',
ERR! Triggers   errno: 1146,
ERR! Triggers   sqlState: '42S02',
ERR! Triggers   index: 0 }
ERR! Feed Error: ER_NO_SUCH_TABLE: Table 'mailtrain_test.campaigns' doesn't exist
ERR! Feed     at Query.Sequence._packetToError (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Sequence.js:52:14)
ERR! Feed     at Query.ErrorPacket (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Query.js:77:18)
ERR! Feed     at Protocol._parsePacket (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:280:23)
ERR! Feed     at Parser.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Parser.js:75:12)
ERR! Feed     at Protocol.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:39:16)
ERR! Feed     at Socket.<anonymous> (/opt/mailtrain/node_modules/mysql/lib/Connection.js:103:28)
ERR! Feed     at emitOne (events.js:96:13)
ERR! Feed     at Socket.emit (events.js:188:7)
ERR! Feed     at readableAddChunk (_stream_readable.js:176:18)
ERR! Feed     at Socket.Readable.push (_stream_readable.js:134:10)
ERR! Feed     --------------------
ERR! Feed     at Protocol._enqueue (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:141:48)
ERR! Feed     at PoolConnection.query (/opt/mailtrain/node_modules/mysql/lib/Connection.js:208:25)
ERR! Feed     at db.getConnection (/opt/mailtrain/services/feedcheck.js:25:20)
ERR! Feed     at Ping.onOperationComplete [as _callback] (/opt/mailtrain/node_modules/mysql/lib/Pool.js:110:5)
ERR! Feed     at Ping.Sequence.end (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Sequence.js:86:24)
ERR! Feed     at Ping.Sequence.OkPacket (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Sequence.js:95:8)
ERR! Feed     at Protocol._parsePacket (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:280:23)
ERR! Feed     at Parser.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Parser.js:75:12)
ERR! Feed     at Protocol.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:39:16)
ERR! Feed     at Socket.<anonymous> (/opt/mailtrain/node_modules/mysql/lib/Connection.js:103:28)
ERR! Feed  { Error: ER_NO_SUCH_TABLE: Table 'mailtrain_test.campaigns' doesn't exist
ERR! Feed     at Query.Sequence._packetToError (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Sequence.js:52:14)
ERR! Feed     at Query.ErrorPacket (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Query.js:77:18)
ERR! Feed     at Protocol._parsePacket (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:280:23)
ERR! Feed     at Parser.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Parser.js:75:12)
ERR! Feed     at Protocol.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:39:16)
ERR! Feed     at Socket.<anonymous> (/opt/mailtrain/node_modules/mysql/lib/Connection.js:103:28)
ERR! Feed     at emitOne (events.js:96:13)
ERR! Feed     at Socket.emit (events.js:188:7)
ERR! Feed     at readableAddChunk (_stream_readable.js:176:18)
ERR! Feed     at Socket.Readable.push (_stream_readable.js:134:10)
ERR! Feed     --------------------
ERR! Feed     at Protocol._enqueue (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:141:48)
ERR! Feed     at PoolConnection.query (/opt/mailtrain/node_modules/mysql/lib/Connection.js:208:25)
ERR! Feed     at db.getConnection (/opt/mailtrain/services/feedcheck.js:25:20)
ERR! Feed     at Ping.onOperationComplete [as _callback] (/opt/mailtrain/node_modules/mysql/lib/Pool.js:110:5)
ERR! Feed     at Ping.Sequence.end (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Sequence.js:86:24)
ERR! Feed     at Ping.Sequence.OkPacket (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Sequence.js:95:8)
ERR! Feed     at Protocol._parsePacket (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:280:23)
ERR! Feed     at Parser.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Parser.js:75:12)
ERR! Feed     at Protocol.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:39:16)
ERR! Feed     at Socket.<anonymous> (/opt/mailtrain/node_modules/mysql/lib/Connection.js:103:28)
ERR! Feed   code: 'ER_NO_SUCH_TABLE',
ERR! Feed   errno: 1146,
ERR! Feed   sqlState: '42S02',
ERR! Feed   index: 0 }
ERR! Import Error: ER_NO_SUCH_TABLE: Table 'mailtrain_test.importer' doesn't exist
ERR! Import     at Query.Sequence._packetToError (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Sequence.js:52:14)
ERR! Import     at Query.ErrorPacket (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Query.js:77:18)
ERR! Import     at Protocol._parsePacket (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:280:23)
ERR! Import     at Parser.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Parser.js:75:12)
ERR! Import     at Protocol.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:39:16)
ERR! Import     at Socket.<anonymous> (/opt/mailtrain/node_modules/mysql/lib/Connection.js:103:28)
ERR! Import     at emitOne (events.js:96:13)
ERR! Import     at Socket.emit (events.js:188:7)
ERR! Import     at readableAddChunk (_stream_readable.js:176:18)
ERR! Import     at Socket.Readable.push (_stream_readable.js:134:10)
ERR! Import     --------------------
ERR! Import     at Protocol._enqueue (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:141:48)
ERR! Import     at PoolConnection.query (/opt/mailtrain/node_modules/mysql/lib/Connection.js:208:25)
ERR! Import     at db.getConnection (/opt/mailtrain/services/importer.js:23:20)
ERR! Import     at Ping.onOperationComplete [as _callback] (/opt/mailtrain/node_modules/mysql/lib/Pool.js:110:5)
ERR! Import     at Ping.Sequence.end (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Sequence.js:86:24)
ERR! Import     at Ping.Sequence.OkPacket (/opt/mailtrain/node_modules/mysql/lib/protocol/sequences/Sequence.js:95:8)
ERR! Import     at Protocol._parsePacket (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:280:23)
ERR! Import     at Parser.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Parser.js:75:12)
ERR! Import     at Protocol.write (/opt/mailtrain/node_modules/mysql/lib/protocol/Protocol.js:39:16)
ERR! Import     at Socket.<anonymous> (/opt/mailtrain/node_modules/mysql/lib/Connection.js:103:28)
```
___

```
[root@mail mailtrain]# npm run e2e

> mailtrain@1.23.2 e2e /opt/mailtrain
> npm run sqlresettest && npm run _e2e


> mailtrain@1.23.2 sqlresettest /opt/mailtrain
> NODE_ENV=test npm run sqldrop && NODE_ENV=test npm run sqlinit


> mailtrain@1.23.2 sqldrop /opt/mailtrain
> node setup/sql/drop.js

info sqldrop Command completed, all tables dropped from "mailtrain_test"

> mailtrain@1.23.2 sqlinit /opt/mailtrain
> node setup/sql/init.js

info sql SQL not set up, initializing
info sql Loading tables from mailtrain-test.sql
info sql Database check completed

> mailtrain@1.23.2 _e2e /opt/mailtrain
> NODE_ENV=test mocha test/e2e/index.js

/opt/mailtrain/test/e2e/tests/login.js:18
    it('can access home page', async () => {
                                     ^
SyntaxError: Unexpected token (
    at createScript (vm.js:56:10)
    at Object.runInThisContext (vm.js:97:10)
    at Module._compile (module.js:542:28)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Suite.describe (/opt/mailtrain/test/e2e/index.js:31:13)
    at Object.create (/opt/mailtrain/node_modules/mocha/lib/interfaces/common.js:114:19)
    at context.describe.context.context (/opt/mailtrain/node_modules/mocha/lib/interfaces/bdd.js:44:27)
    at tests.forEach.t (/opt/mailtrain/test/e2e/index.js:30:9)
    at Array.forEach (native)
    at Suite.<anonymous> (/opt/mailtrain/test/e2e/index.js:29:11)
    at Object.create (/opt/mailtrain/node_modules/mocha/lib/interfaces/common.js:114:19)
    at context.describe.context.context (/opt/mailtrain/node_modules/mocha/lib/interfaces/bdd.js:44:27)
    at Object.<anonymous> (/opt/mailtrain/test/e2e/index.js:26:1)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at /opt/mailtrain/node_modules/mocha/lib/mocha.js:230:27
    at Array.forEach (native)
    at Mocha.loadFiles (/opt/mailtrain/node_modules/mocha/lib/mocha.js:227:14)
    at Mocha.run (/opt/mailtrain/node_modules/mocha/lib/mocha.js:495:10)
    at Object.<anonymous> (/opt/mailtrain/node_modules/mocha/bin/_mocha:469:18)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3)
    at Module.runMain (module.js:604:10)
    at run (bootstrap_node.js:393:7)
    at startup (bootstrap_node.js:150:9)
    at bootstrap_node.js:508:3

npm ERR! Linux 3.10.0-514.16.1.el7.x86_64
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "run" "_e2e"
npm ERR! node v6.10.2
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! mailtrain@1.23.2 _e2e: `NODE_ENV=test mocha test/e2e/index.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the mailtrain@1.23.2 _e2e script 'NODE_ENV=test mocha test/e2e/index.js'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the mailtrain package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     NODE_ENV=test mocha test/e2e/index.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs mailtrain
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls mailtrain
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /opt/mailtrain/npm-debug.log

npm ERR! Linux 3.10.0-514.16.1.el7.x86_64
npm ERR! argv "/usr/bin/node" "/usr/bin/npm" "run" "e2e"
npm ERR! node v6.10.2
npm ERR! npm  v3.10.10
npm ERR! code ELIFECYCLE
npm ERR! mailtrain@1.23.2 e2e: `npm run sqlresettest && npm run _e2e`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the mailtrain@1.23.2 e2e script 'npm run sqlresettest && npm run _e2e'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the mailtrain package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     npm run sqlresettest && npm run _e2e
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs mailtrain
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls mailtrain
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /opt/mailtrain/npm-debug.log
```

**Comment by witzig** [comment]
The errors Mailtrain is throwing are 'normal', caused by resetting the database while the app is running.

The error for e2e tests:

```
/opt/mailtrain/test/e2e/tests/login.js:18
    it('can access home page', async () => {

SyntaxError: Unexpected token (
```

Seems your Node version doesn't support async/await. Are you running 7.6 or later?

File: test/e2e/README.md (hunk 1 above)
Comment (witzig) [inline test/e2e/README.md]: Thanks.

**Comment by bures** [comment]
Good catch. I though I had the newest stable version on NodeJS, but it seems I have the LTS one.

**Comment by witzig** [comment]
Sure hope this solves the problem. I went with async/await for readability and although it's not nice to require Node 7.6 (or later) for the tests, it's certainly an issue time itself will solve.

Also the errors Mailtrain is throwing when resetting the database ... that's not nice. Once the DB is set up the current tests don't actually need a reset, thus running the tests with `npm run _e2e` is fine.

I guess we should probably add an additional command:
1. reset the DB
2. start Mailtrain
3. run the tests
4. kill Mailtrain

**Comment by bures** [comment]
Thanks a lot. Actually I like async/await. It makes the code so much more readable. I was even wondering whether we should not move the main codebase to Node 7 and use async/await.
T

**Comment by witzig** [comment]
> I was even wondering whether we should not move the main codebase to Node 7 and use async/await.

Well, no objections from my side ... 

I probably take a stab at Grapejs next. A true MJML WYSWNG (and code) editor would be awesome.

**Comment by andris9** [comment]
I'd also support using Node v7. There are a lot of users who probably can't upgrade right now but eventually they would be.

**Comment by bures** [comment]
So do we have consensus that from now on we switch to Node v7 to develop new features?
I'm definitely in support of it.

**Comment by witzig** [comment]
It looks like, yes. For new features that definitely makes the most sense. 

@andris9, how do you want to handle this? Using a `next` branch?

**Comment by andris9** [comment]
I'm not sure about branching. I'd rather tag the last state of the code that runs on Node v6 as v1.24 and the first version that requires Node 7 as v2.0.

**Comment by witzig** [comment]
Perfect! So, let's tag master? Then merge this PR (which already requires v7) after #227.

**Comment by andris9** [comment]
I tagged the current master and added a note to README. You can now merge PRs that use Node.js v7

**Comment by witzig** [comment]
Thanks!

**Comment by bures** [comment]
Perfect.

**Comment by bures** [comment]
Just tried it. Works lovely.

@andris9 - As far as I am concerned, it can be merged in. Do you want to do it yourself or shall I do it?

**Comment by andris9** [comment]
@bures it's fine by me if you do everything related tests etc yourself

**Pull request merged**